src/super_gradients/conversion/onnx/nms.py

```diff
@@ -32,49 +44,78 @@ def __init__(self, batch_size: int, num_pre_nms_predictions: int, max_prediction
         selected_scores = pred_scores[batch_indexes, boxes_indexes, label_indexes]
 
-        predictions = torch.cat([batch_indexes.unsqueeze(1), selected_boxes, selected_scores.unsqueeze(1), label_indexes.unsqueeze(1)], dim=1)
+        if self.batch_size == 1:
+            pred_boxes = selected_boxes[: self.max_predictions_per_image]
+            pred_scores = selected_scores[: self.max_predictions_per_image]
+            pred_classes = label_indexes[: self.max_predictions_per_image].long()
+            num_predictions = pred_boxes.size(0).reshape(1, 1)
 
-        predictions = torch.nn.functional.pad(
-            predictions, (0, 0, 0, self.max_predictions_per_image * self.batch_size - predictions.size(0)), value=-1, mode="constant"
-        )
+            pad_size = self.max_predictions_per_image - pred_boxes.size(0)
+            pred_boxes = torch.nn.functional.pad(pred_boxes, [0, 0, 0, pad_size], value=-1, mode="constant")
+            pred_scores = torch.nn.functional.pad(pred_scores, [0, pad_size], value=-1, mode="constant")
+            pred_classes = torch.nn.functional.pad(pred_classes, [0, pad_size], value=-1, mode="constant")
 
-        batch_predictions = torch.zeros((self.batch_size, self.max_predictions_per_image, 6), dtype=predictions.dtype, device=predictions.device)
+            return num_predictions, pred_boxes.unsqueeze(0), pred_scores.unsqueeze(0), pred_classes.unsqueeze(0)
+        else:
+            predictions = torch.cat([selected_boxes, selected_scores.unsqueeze(1), label_indexes.unsqueeze(1)], dim=1)
 
-        batch_indexes = torch.arange(start=0, end=self.batch_size, step=1, device=predictions.device).to(dtype=predictions.dtype)
-        masks = batch_indexes.view(-1, 1).eq(predictions[:, 0].view(1, -1))  # [B, N]
+            batch_predictions = torch.zeros((self.batch_size, self.max_predictions_per_image, 6), dtype=predictions.dtype, device=predictions.device)
 
-        num_predictions = torch.sum(masks, dim=1).long()
+            image_indexes = torch.arange(start=0, end=self.batch_size, step=1, device=predictions.device)
+            masks = image_indexes.view(self.batch_size, 1) == batch_indexes.view(1, selected_indexes.size(0))  # [B, L]
 
-        for i in range(self.batch_size):
-            selected_predictions = predictions[masks[i]]
-            selected_predictions = selected_predictions[:, 1:]
-            batch_predictions[i] = torch.nn.functional.pad(
-                selected_predictions, (0, 0, 0, self.max_predictions_per_image - selected_predictions.size(0)), value=0, mode="constant"
-            )
+            # Add dummy row to mask and predictions to ensure that we always have at least one prediction per image
+            # ONNX/TRT deals poorly with tensors that has zero dims, and we need to ensure that we always have at least one prediction per image
+            masks = torch.cat([masks, torch.zeros((self.batch_size, 1), dtype=masks.dtype, device=predictions.device)], dim=1)  # [B, L+1]
```

Comment:
The problem we are facing with TRT is some sort of edge case when you have tensor of shape `[Batch, Num Detections]`. when `Num Detections = 0`.
Empty predictions may appear when there is no detections in batch. 

And for some reason even basic constructions (`F.pad` or `view(-1)` and `squeeze(1)` do not work properly. Even `torch.sum` fails when input has `[B, 0]` shape). The only workaround I was able to find to combat this scenario is `torch.cat`. 

So the intention is indeed correct - adding a False/Zero items to ensure that shape of masks/preds has at least one element in Num Detections. Since mask padded with zeros - it does not affect existing predictions. 

In the nutshell - this is a dirty trick to fool TRT and make it do what we want him to do 🙌

Reply:
I understand.
But in the case where num_detections=0, we add dummy row and another dummy row to the masks = False.
Then we won't select the dummy row, and still end up with a size=0 in the last dimension.
Or is it ok from that point on?

Maybe I missed it - but I didn't see we test export works in the case of no predictions.
Can we add such?

Reply:
Gather by mask when mask has shape [B,0] is somehow works fine. It is mainly for num_predictions = torch.sum(masks, dim=1, keepdim=True).long()